src/connect/planner.cpp

```diff
@@ -205,25 +205,35 @@ void Planner::command(const Command &command, const Gcode &gcode) {
     planned_event = Event { EventType::Accepted, command.id };
 }
 
-#define JC(CMD)                                                         \
-    void Planner::command(const Command &command, const CMD##Print &) { \
-        if (printer.job_control(Printer::JobControl::CMD)) {            \
-            planned_event = Event { EventType::Finished, command.id };  \
-        } else {                                                        \
-            planned_event = Event { EventType::Rejected, command.id };  \
-        }                                                               \
+#define JC(CMD, REASON)                                                                          \
+    void Planner::command(const Command &command, const CMD##Print &) {                          \
+        if (printer.job_control(Printer::JobControl::CMD)) {                                     \
+            planned_event = Event { EventType::Finished, command.id };                           \
+        } else {                                                                                 \
+            planned_event = Event { EventType::Rejected, command.id, nullopt, nullopt, REASON }; \
+        }                                                                                        \
     }
 
-JC(Pause)
-JC(Resume)
-JC(Stop)
+JC(Pause, "No print to pause")
+JC(Resume, "No paused print to resume")
+JC(Stop, "No print to stop")
 
 void Planner::command(const Command &command, const StartPrint &params) {
     const char *path = params.path.path();
-    if (path_allowed(path) && path_exists(path) && printer.start_print(path)) {
+
+    const char *reason = nullptr;
+    if (!path_allowed(path)) {
+        reason = "Forbidden path";
+    } else if (!path_exists(path)) {
+        reason = "File not found";
+    } else if (!printer.start_print(path)) {
+        reason = "Can't print now";
+    }
+
+    if (reason == nullptr) {
```

Comment:
Generally I think checking for `nullptr` directly with `if (reason == nullptr)` => `if (!reason)` and vice versa is cleaner. Just like one doesn't write `if (boolean_flag == false)` but `if (!boolean_flag)`.